service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/disk/DetachAzureDiskFromVmStep.java

```diff
@@ -0,0 +1,196 @@
+      }
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    final AzureCloudContext azureCloudContext =
+        DeleteAzureDiskFlightUtils.getAzureCloudContext(context);
+    final String diskResourceId =
+        DeleteAzureDiskFlightUtils.getAzureDiskResourceId(azureCloudContext, resource);
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+    try {
+      var disk = computeManager.disks().getById(diskResourceId);
+      var vmId =
+          Optional.ofNullable(
+              context
+                  .getWorkingMap()
+                  .get(DeleteAzureDiskFlightUtils.DISK_ATTACHED_VM_ID_KEY, String.class));
+      if (vmId.isEmpty()) {
+        // noop in case disk wasn't initially attached to a virtual machine
+        return StepResult.getStepResultSuccess();
+      }
+      var vm = getVirtualMachine(computeManager, vmId.get());
+      vm.ifPresent(virtualMachine -> attachDisk(virtualMachine, disk));
+    } catch (ManagementException ex) {
+      if (AzureManagementExceptionUtils.isExceptionCode(
+          ex, AzureManagementExceptionUtils.RESOURCE_NOT_FOUND)) {
+        logger.info(
+            "Disk {} does not exist in MRG {} for workspace {}.",
+            resource.getResourceId(),
+            azureCloudContext.getAzureResourceGroupId(),
+            resource.getWorkspaceId());
+        return StepResult.getStepResultSuccess();
+      }
+      if (AzureManagementExceptionUtils.isExceptionCode(
+          ex, AzureManagementExceptionUtils.OPERATION_NOT_ALLOWED)) {
+        logger.error(
+            "Failed attempt to attach disk {} back to vm in MRG {} for workspace {}.",
+            resource.getResourceId(),
+            azureCloudContext.getAzureResourceGroupId(),
+            resource.getWorkspaceId());
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, ex);
+      }
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  private Optional<VirtualMachine> getVirtualMachine(ComputeManager computeManager, String vmId) {
+    VirtualMachine vm;
+    try {
+      vm = computeManager.virtualMachines().getById(vmId);
+    } catch (ManagementException ex) {
+      if (AzureManagementExceptionUtils.isExceptionCode(
+          ex, AzureManagementExceptionUtils.RESOURCE_NOT_FOUND)) {
+        logger.info(
+            "Vm to which disk was attached does not exist in workspace {}.",
+            resource.getWorkspaceId());
+        return Optional.empty();
+      }
+      throw ex;
+    }
+    return Optional.of(vm);
+  }
+
+  private void detachDisk(VirtualMachine vm, Integer lun) {
+    var vmUpdate = vm.update();
+    vmUpdate.withoutDataDisk(lun);
```

Comment:
This methods don't return the mutated state? I'm accustomed to seeing chaining, like

`vm.update().withoutDataDisk(lun).apply()`

Reply:
That's valid observation. Current implementation is mutating its state. Here is vm [implementation](https://github.com/Azure/azure-libraries-for-java/blob/22f90e7c258c304eafe25c2558ddf7de81563198/azure-mgmt-compute/src/main/java/com/microsoft/azure/management/compute/implementation/VirtualMachineImpl.java#L1096). It is not necessary, but I will update it.